File: api/src/main/java/org/pmiops/workbench/audit/AbstractAuditableEvent.java
package org.pmiops.workbench.audit;

import static org.pmiops.workbench.api.StatusController.LOGGING_RESOURCE;

import com.google.cloud.logging.LogEntry;
import com.google.cloud.logging.Payload.JsonPayload;
import com.google.cloud.logging.Severity;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public abstract class AbstractAuditableEvent {

  private static final String LOG_NAME = "action-audit-test-2";

  public abstract long timestamp();

  public abstract String actionId();

  public abstract AgentType agentType();

  public abstract long agentId();

  public abstract Optional<String> agentEmail();

  public abstract ActionType actionType();

  public abstract TargetType targetType();

  public abstract Optional<Long> targetId();

  public abstract Optional<String> targetProperty();

  public abstract Optional<String> previousValue();

  public abstract Optional<String> newValue();

  private JsonPayload toJsonPayload() {
    Map<String, Object> result = new HashMap<>();
    result.put(AuditColumn.TIMESTAMP.name(), timestamp());
    result.put(AuditColumn.ACTION_ID.name(), actionId());
    result.put(AuditColumn.ACTION_TYPE.name(), actionType());
    result.put(AuditColumn.AGENT_TYPE.name(), agentType());
    result.put(AuditColumn.AGENT_ID.name(), agentId());
    result.put(AuditColumn.AGENT_EMAIL.name(), agentEmail().orElse(null));
    result.put(AuditColumn.TARGET_TYPE.name(), targetType());
    result.put(AuditColumn.TARGET_ID.name(), targetId().orElse(null));
    result.put(AuditColumn.AGENT_ID.name(), targetId().orElse(null));
    result.put(AuditColumn.TARGET_PROPERTY.name(), targetProperty().orElse(null));
    result.put(AuditColumn.PREV_VALUE.name(), previousValue().orElse(null));
    result.put(AuditColumn.NEW_VALUE.name(), newValue().orElse(null));
    return JsonPayload.of(result);
  }

  public LogEntry toLogEntry() {
    return LogEntry.newBuilder(toJsonPayload())
        .setSeverity(Severity.INFO)
        .setLogName(LOG_NAME)
        .setResource(LOGGING_RESOURCE)
        .build();
  }
}

File: api/src/main/java/org/pmiops/workbench/audit/AbstractAuditableEvent.kt
package org.pmiops.workbench.audit

import com.google.cloud.MonitoredResource
import com.google.cloud.logging.LogEntry
import com.google.cloud.logging.Payload.JsonPayload
import com.google.cloud.logging.Severity
import java.util.HashMap
import java.util.Optional

abstract class AbstractAuditableEvent {

    abstract fun timestamp(): Long

    abstract fun actionId(): String

    abstract fun agentType(): AgentType

    abstract fun agentId(): Long

    abstract fun agentEmail(): Optional<String>

    abstract fun actionType(): ActionType

    abstract fun targetType(): TargetType

    abstract fun targetId(): Optional<Long>

    abstract fun targetProperty(): Optional<String>

    abstract fun previousValue(): Optional<String>

    abstract fun newValue(): Optional<String>

    private fun toJsonPayload(): JsonPayload {
        val result = HashMap<String, Any>() // allow null values
        result[AuditColumn.TIMESTAMP.name] = timestamp()
        result[AuditColumn.ACTION_ID.name] = actionId()
        result[AuditColumn.ACTION_TYPE.name] = actionType()
        result[AuditColumn.AGENT_TYPE.name] = agentType()
        result[AuditColumn.AGENT_ID.name] = agentId()
        result[AuditColumn.AGENT_EMAIL.name] = toNullable(agentEmail())
        result[AuditColumn.TARGET_TYPE.name] = targetType()
        result[AuditColumn.TARGET_ID.name] = toNullable(targetId())
        result[AuditColumn.AGENT_ID.name] = agentId()
        result[AuditColumn.TARGET_PROPERTY.name] = toNullable(targetProperty())
        result[AuditColumn.PREV_VALUE.name] = toNullable(previousValue())
        result[AuditColumn.NEW_VALUE.name] = toNullable(newValue())
        return JsonPayload.of(result)
    }

    fun toLogEntry(): LogEntry {

        return LogEntry.newBuilder(toJsonPayload())
                .setSeverity(Severity.INFO)
                .setLogName(LOG_NAME)
                .setResource(MONITORED_RESOURCE)
                .build()
    }

    companion object {
        private val MONITORED_RESOURCE = MonitoredResource.newBuilder("global").build()
        private val LOG_NAME = "action-audit-test-3"

        // Inverse of Optional.ofNullable(). Used for JSON api which expects null values for empty
        // columns. Though perhaps we could just leave those out...
        private fun <T> toNullable(opt: Optional<T>): T {
            return opt.orElse(null)
        }
    }
}
